File: content/snippets/cpp/design_patterns_chain_of_responsibility.cpp
#include <iostream>
#include <memory>
#include <string>

class Request {
public:
    int level;
    std::string message;
    Request(int l, const std::string& m) : level(l), message(m) {}
};

class Handler {
protected:
    std::shared_ptr<Handler> next;
public:
    void setNext(std::shared_ptr<Handler> handler) { next = handler; }
    virtual void handle(const Request& req) = 0;
};

class LowLevelHandler : public Handler {
public:
    void handle(const Request& req) override {
        if (req.level <= 1) {
            std::cout << "LowLevelHandler: Processing " << req.message << std::endl;
        } else if (next) {
            next->handle(req);
        }
    }
};

class MidLevelHandler : public Handler {
public:
    void handle(const Request& req) override {
        if (req.level == 2) {
            std::cout << "MidLevelHandler: Processing " << req.message << std::endl;
        } else if (next) {
            next->handle(req);
        }
    }
};

class HighLevelHandler : public Handler {
public:
    void handle(const Request& req) override {
        if (req.level >= 3) {
            std::cout << "HighLevelHandler: Processing " << req.message << std::endl;
        }
    }
};

int main() {
    auto low = std::make_shared<LowLevelHandler>();
    auto mid = std::make_shared<MidLevelHandler>();
    auto high = std::make_shared<HighLevelHandler>();
    
    low->setNext(mid);
    mid->setNext(high);
    
    Request r1(1, "Simple request");
    Request r2(2, "Normal request");
    Request r3(3, "Critical request");
    
    low->handle(r1);
    low->handle(r2);
    low->handle(r3);
}

File: content/snippets/go/design_patterns_chain_of_responsibility.go
package main

import (
	"fmt"
)

type Request struct {
	level   int
	message string
}

type Handler interface {
	SetNext(Handler)
	Handle(*Request)
}

type AbstractHandler struct {
	next Handler
}

func (h *AbstractHandler) SetNext(next Handler) {
	h.next = next
}

type LowHandler struct {
	AbstractHandler
}

func (h *LowHandler) Handle(req *Request) {
	if req.level <= 1 {
		fmt.Printf("LowHandler: Processing %s\n", req.message)
	} else if h.next != nil {
		h.next.Handle(req)
	}
}

type MidHandler struct {
	AbstractHandler
}

func (h *MidHandler) Handle(req *Request) {
	if req.level == 2 {
		fmt.Printf("MidHandler: Processing %s\n", req.message)
	} else if h.next != nil {
		h.next.Handle(req)
	}
}

type HighHandler struct {
	AbstractHandler
}

func (h *HighHandler) Handle(req *Request) {
	if req.level >= 3 {
		fmt.Printf("HighHandler: Processing %s\n", req.message)
	}
}

func main() {
	low := &LowHandler{}
	mid := &MidHandler{}
	high := &HighHandler{}

	low.SetNext(mid)
	mid.SetNext(high)

	low.Handle(&Request{1, "Simple"})
	low.Handle(&Request{2, "Normal"})
	low.Handle(&Request{3, "Critical"})
}
